ledger/acctonline.go

```diff
@@ -0,0 +1,846 @@
+	// for the range [Lastest - MaxBalLookback - X, Latest - lookback], where X = [0, commit range]
+	// and alway containing an entry for Lastest - MaxBalLookback + 1 if some account is cached.
+	// The invariant is held by
+	// 1) loading a full history when new accounts get added
+	// 2) adding online accounts state changes when flushing to disk
+	// 3) pruning the history by removing older than Lastest - MaxBalLookback non-online entries
+	onlineAccountsCache onlineAccountsCache
+
+	// maxAcctLookback sets the minimim deltas size to keep in memory
+	acctLookback uint64
+}
+
+// initialize initializes the accountUpdates structure
+func (ao *onlineAccounts) initialize(cfg config.Local) {
+	ao.accountsReadCond = sync.NewCond(ao.accountsMu.RLocker())
+	ao.acctLookback = cfg.MaxAcctLookback
+}
+
+// loadFromDisk is the 2nd level initialization, and is required before the accountUpdates becomes functional
+// The close function is expected to be call in pair with loadFromDisk
+func (ao *onlineAccounts) loadFromDisk(l ledgerForTracker, lastBalancesRound basics.Round) error {
+	ao.accountsMu.Lock()
+	defer ao.accountsMu.Unlock()
+
+	ao.cachedDBRoundOnline = lastBalancesRound
+	ao.ledger = l
+	err := ao.initializeFromDisk(l, lastBalancesRound)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// initializeFromDisk performs the atomic operation of loading the accounts data information from disk
+// and preparing the accountUpdates for operation.
+func (ao *onlineAccounts) initializeFromDisk(l ledgerForTracker, lastBalancesRound basics.Round) (err error) {
+	ao.dbs = l.trackerDB()
+	ao.log = l.trackerLog()
+
+	err = ao.dbs.Wdb.Atomic(func(ctx context.Context, tx *sql.Tx) error {
+		var err0 error
+		var endRound basics.Round
+		ao.onlineRoundParamsData, endRound, err0 = accountsOnlineRoundParams(tx)
+		if err0 != nil {
+			return err0
+		}
+		if endRound != ao.cachedDBRoundOnline {
+			return fmt.Errorf("last onlineroundparams round %d does not match dbround %d", endRound, ao.cachedDBRoundOnline)
+		}
+
+		onlineAccounts, err0 := onlineAccountsAll(tx, onlineAccountsCacheMaxSize)
+		if err0 != nil {
+			return err0
+		}
+		ao.onlineAccountsCache.init(onlineAccounts, onlineAccountsCacheMaxSize)
+
+		return nil
+	})
+	if err != nil {
+		return
+	}
+
+	ao.accountsq, err = onlineAccountsInitDbQueries(ao.dbs.Rdb.Handle)
+	if err != nil {
+		return
+	}
+
+	ao.deltas = nil
+	ao.accounts = make(map[basics.Address]modifiedOnlineAccount)
+	ao.deltasAccum = []int{0}
+
+	ao.baseOnlineAccounts.init(ao.log, baseAccountsPendingAccountsBufferSize, baseAccountsPendingAccountsWarnThreshold)
+	return
+}
+
+// latest returns the latest round
+func (ao *onlineAccounts) latest() basics.Round {
+	return ao.cachedDBRoundOnline + basics.Round(len(ao.deltas))
+}
+
+// close closes the accountUpdates, waiting for all the child go-routine to complete
+func (ao *onlineAccounts) close() {
```

Comment:
is the mutex required here?

Reply:
we called this under trackersMu from above